l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol

```diff
@@ -137,20 +149,68 @@ abstract contract BaseZkSyncUpgrade is ZKChainBase {
         emit NewL2BootloaderBytecodeHash(previousBootloaderBytecodeHash, _l2BootloaderBytecodeHash);
     }
 
-    /// @notice Change the address of the verifier smart contract
-    /// @param _newVerifier Verifier smart contract address
-    function _setVerifier(IVerifier _newVerifier) private {
+    /// @notice Change the address of the Wrapper verifier smart contract
+    /// @param _newDualVerifier DualVerifier smart contract address
+    function _setDualVerifier(IVerifier _newDualVerifier) private {
+        // An upgrade to the verifier must be done carefully to ensure there aren't batches in the committed state
+        // during the transition. If verifier is upgraded, it will immediately be used to prove all committed batches.
+        // Batches committed expecting the old verifier will fail. Ensure all committed batches are finalized before the
+        // verifier is upgraded.
+        if (_newDualVerifier == IVerifier(address(0))) {
+            return;
+        }
+
+        IVerifier oldDualVerifier = s.dualVerifier;
+        s.dualVerifier = _newDualVerifier;
+        emit NewDualVerifier(address(oldDualVerifier), address(_newDualVerifier));
+    }
+
+    /// @notice Change the address of the PLONK verifier smart contract
+    /// @param _newPlonkVerifier PlonkVerifier smart contract address
+    function _setPlonkVerifier(address _newPlonkVerifier) private {
+        // An upgrade to the verifier must be done carefully to ensure there aren't batches in the committed state
+        // during the transition. If verifier is upgraded, it will immediately be used to prove all committed batches.
+        // Batches committed expecting the old verifier will fail. Ensure all committed batches are finalized before the
+        // verifier is upgraded.
+        if (_newPlonkVerifier == address(0)) {
+            return;
+        }
+
+        address oldPlonkVerifier = s.plonkVerifier;
+        s.plonkVerifier = _newPlonkVerifier;
+        emit NewPlonkVerifier(oldPlonkVerifier, _newPlonkVerifier);
+    }
+
+    /// @notice Change the address of the FFLONK verifier smart contract
+    /// @param _newFflonkVerifier FflonkVerifier smart contract address
+    function _setFflonkVerifier(address _newFflonkVerifier) private {
```

Comment:
system should work only with one verifier, no need to set fflonk and plonk separately. The only thing zk chain should care is the address of the verifier, dual verifier should route the verification to one of them